commands/auto/followtrajectory.py

```diff
@@ -0,0 +1,213 @@
+            for state in self.originTrajectory.getStates()
+        ]
+
+        SmartDashboard.putNumberArray(
+            constants.kAutonomousPathKey,
+            functools.reduce(
+                operator.add,
+                [
+                    [state[0].X(), state[0].Y(), state[1].radians()]
+                    for state in transformedStates
+                ],
+                [],
+            ),
+        )
+        DataLogManager.log("begin trajectory")
+
+    def getAllianceRespectivePoint(
+        self, pose: Pose2d, holonomicRotation: Rotation2d
+    ) -> Tuple[Pose2d, Rotation2d]:
+        if DriverStation.getAlliance() == DriverStation.Alliance.kRed:
+            return (
+                Pose2d(
+                    constants.kFieldLength - pose.X(),
+                    pose.Y(),
+                    -pose.rotation() + Rotation2d(pi),
+                ),
+                -holonomicRotation + Rotation2d(pi),
+            )
+        else:
+            return (pose, holonomicRotation)
+
+    def execute(self) -> None:
+        for command, running in self.currentCommands:
+            if not running:
+                continue
+            command.execute()
+
+            if command.isFinished():
+                command.end(False)
+                running = False
+
+        curTime = self.timer.get()
+        currentState = self.drive.getPose()
+        desiredState = self.trajectory.sample(curTime)
+
+        allianceRespectiveDesiredState = self.getAllianceRespectivePoint(
+            desiredState.pose, desiredState.holonomicRotation
+        )
+
+        SmartDashboard.putNumberArray(
+            constants.kAutonomousPathError,
+            [
+                currentState.X() - allianceRespectiveDesiredState[0].X(),
+                currentState.Y() - allianceRespectiveDesiredState[0].Y(),
+                (currentState.rotation() - allianceRespectiveDesiredState[1]).radians(),
+            ],
+        )
+        SmartDashboard.putNumberArray(
+            constants.kAutonomousPathSample,
+            [
+                allianceRespectiveDesiredState[0].X(),
+                allianceRespectiveDesiredState[0].Y(),
+                allianceRespectiveDesiredState[1].radians(),
+            ],
+        )
+
+        targetChassisSpeeds = self.controller.calculate(
+            currentState,
+            allianceRespectiveDesiredState[0],
+            desiredState.velocity,
+            allianceRespectiveDesiredState[1],
+        )
+
+        SmartDashboard.putNumberArray(
+            constants.kAutonomousChassisSpeeds,
+            [targetChassisSpeeds.vx, targetChassisSpeeds.vy, targetChassisSpeeds.omega],
+        )
+
+        self.drive.arcadeDriveWithSpeeds(
+            targetChassisSpeeds, DriveSubsystem.CoordinateMode.RobotRelative
+        )
+
+        # marker relatev stuff
+        if len(self.unpassedMarkers) > 0 and curTime >= self.unpassedMarkers[0].time:
```

Comment:
Is the intention here that if we exceed this time we move on to the next marker regardless of where we are?  Or am I misunderstanding that?

Reply:
the idea is that if there are markers on the path we havn't hit yet, and we are at or passed the next marker's time to then pop the marker at the front and interpret what to do with it

Reply:
So its a way to "catch up", but conceivably you run the risk of hitting the charge station for example?

Reply:
its the method for how to deal with when we finally hit a markes
the >= ensures that if we are slightly off we do in fact interpret the marker

this is also converted from the pathplannerlib located at https://github.com/mjansen4857/pathplanner/blob/e3a983b83fc267df6cea6b7a2cafa0498e7fb907/pathplannerlib/src/main/java/com/pathplanner/lib/commands/FollowPathWithEvents.java#L91